quinn-proto/src/congestion/bbr_min_max.rs

```diff
@@ -0,0 +1,173 @@
+/*
+ * Based on Google code released under BSD license here:
+ * https://groups.google.com/forum/#!topic/bbr-dev/3RTgkzi5ZD8
+ */
+
+/*
+ * Kathleen Nichols' algorithm for tracking the minimum (or maximum)
+ * value of a data stream over some fixed time interval.  (E.g.,
+ * the minimum RTT over the past five minutes.) It uses constant
+ * space and constant time per update yet almost always delivers
+ * the same minimum as an implementation that has to keep all the
+ * data in the window.
+ *
+ * The algorithm keeps track of the best, 2nd best & 3rd best min
+ * values, maintaining an invariant that the measurement time of
+ * the n'th best >= n-1'th best. It also makes sure that the three
+ * values are widely separated in the time window since that bounds
+ * the worse case error when that data is monotonically increasing
+ * over the window.
+ *
+ * Upon getting a new min, we can forget everything earlier because
+ * it has no value - the new min is <= everything else in the window
+ * by definition and it samples the most recent. So we restart fresh on
+ * every new min and overwrites 2nd & 3rd choices. The same property
+ * holds for 2nd & 3rd best.
+ */
+
+use std::fmt::Debug;
+
+#[derive(Debug, Copy, Clone, Default)]
+struct MinMaxSample {
+    /// round number, not a timestamp
+    time: u64,
+    value: u64,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct MinMax {
+    /// round count, not a timestamp
+    window: u64,
+    samples: [MinMaxSample; 3],
+}
+
+impl MinMax {
+    pub fn new(round_window: u64) -> Self {
+        MinMax {
+            window: round_window,
+            samples: [Default::default(); 3],
+        }
+    }
+
+    pub fn get(&self) -> u64 {
+        self.samples[0].value
+    }
+
+    fn fill(&mut self, sample: MinMaxSample) {
+        self.samples.fill(sample);
+    }
+
+    pub fn reset(&mut self) {
+        self.fill(Default::default())
+    }
+
+    /// unused, but still translated from original source for completeness
```

Comment:
would need an `#[allow(dead_code)]`.  But keeping unused code around is usually an antipattern

Reply:
you'd prefer to remove it then? I'm not attached.

Reply:
I'd prefer to drop it, since otherwise it's at high risk of bitrot. Maybe leave a comment on `update_max` that an inverse version can be defined, and cite a reference for it.

Reply:
done